### ISSUE TITLE
Support for embedded class or object

### ISSUE THREAD

**Issue by mingchuno** [alexeyraspopov/dataclass#6]
It seems that this lib currently doesn't support embedded dataclass or JS object inside the dataclass. Try to add the following to the test.

```
  class Embedded extends Record {
    name: string = 'name';
    age: number = 1;
    entity: Entity;
    date: Date;
  }
  it.only('should compare dataclass with embedded class', () => {
    let embeddedA = new Embedded({
      entity: new Entity(),
      date: new Date('1996-12-17T03:24:00')
    })
    let embeddedB = new Embedded({
      entity: new Entity({ someString: 'sadasd' }),
      date: new Date('1995-12-17T03:24:00')
    })
    console.log(embeddedA.toJSON())
    console.log(embeddedA.entity)
    expect(embeddedA.equals(embeddedB)).toBe(false);
  })
```

I suppose `embeddedA.toJSON()` should also print the `entity` and `date` and `equals` should be false when comparing `embeddedA` and `embeddedB`. Is it possible to add this feature? 
It seems that currently we can only define primitive type inside the dataclass which is a big limitation for the lib to be useful....

**Comment by mingchuno** [comment]
After more details looking, I have found 3 issues:

1. `toJSON()` do not support embedded dataclass
2. `equals()` do not support non-primitive value
3. dataclass definition must supply default value

I have fix 1 and created a PR #7. I cannot fix 2 but I leave a skipped test on it. 3 should be documented some where.

**Comment by alexeyraspopov** [comment]
#7 is merged. Going to make a change for `equals()`.

**Comment by alexeyraspopov** [comment]
in https://github.com/alexeyraspopov/dataclass/commit/4064049969df8859d0e17df2ed093535283e1f9b I made `equals()` checking for nested value objects.

**Comment by alexeyraspopov** [comment]
I'm not sure though about `Date`. It seems to be a special case. For custom objects as value object properties it would be recommended to be converted to value objects, if comparison by value is needed. As for things like `Date` (and, I guess, `Array`) it may require too much recursive logic that can end up being not ideal assumptions about how value comparison must work for those non-primitive native types.

**Comment by alexeyraspopov** [comment]
@mingchuno, I just found the solution that supports use of `Date` in records. In https://github.com/alexeyraspopov/dataclass/commit/8c0cf12c4441864654e93c7bf8307b568215fd4f I extended equality check logic to use `.valueOf()` method whenever it is available. In case of dates, `.valueOf()` returns numeric representation of a date that we can compare by value.

I published v1.2.0 with introduced change.
